### DISCUSSION TITLE
Proposal: 'safe' context to undo 'unsafe' from outer scope?

### DISCUSSION THREAD

**Discussion by fbrosseau** [dotnet/csharplang#6668] [Language Ideas]
Hello,

I tried to search and could not find a similar proposal. I apologize if this is a duplicate.

Say there is a large `unsafe` context (either a large method, or even an entire class). It is currently forbidden to declare an async lambda within this context: `CS4004: Cannot await in an unsafe context`. 

Would it be possible for the language to add a `safe` keyword (or anything similar that would achieve the same goal) that would allow for nested lambdas to use all regular language features? Basically, reverting the `unsafe` context within this lambda.

Concretely, 

```
unsafe void MyUnsafeFunction()
{
    Action myAsyncAction = async () =>
    {
        await Task.Yield();
    }
}
```

The following example is currently forbidden. Could there be any new construct to allow for myAsyncAction to legally use async code?

Or, better yet (but I do not know if this is possible at all :)), could lambdas be _implicitly safe_ so long as they use no unsafe features, and unsafe if they do? This would retain backward compatibility, but as I said I have no idea if this is even possible at the language level, I am just thinking out loud.

Of course, a possible workaround is to not use lambdas and use another standalone method that is in regular non-unsafe context, but then you lose all the power of lambdas, primarily closures.

**Comment by HaloFour** [comment]
> Say there is a large unsafe context (either a large method, or even an entire class).

Why not use `unsafe` blocks instead to limit how much code is considered unsafe?

> Of course, a possible workaround is to not use lambdas and use another standalone method that is in regular non-unsafe context, but then you lose all the power of lambdas, primarily closures.

Which is why it's disallowed.  You're trying to close over an unsafe context.  If you use a `static` lambda, which disables closures, then those lambdas may be declared within an `unsafe` context.
